src/odemis/driver/smaract.py

```diff
@@ -2930,6 +2950,76 @@ def updateMetadata(self, md):
                 raise ValueError("Invalid metadata, should be a coordinate dictionary but got %s." % (deactive_pos,))
         super(MCS2, self).updateMetadata(md)
 
+    # Methods below are for using and applying additional parameters through a param_file
+    def apply_config(self, prop_params):
+        """
+        Configure the device according to the given 'user configuration' from a specified param_file.
+        :param prop_params (dict (int, int, str) -> int/str): channel index, property code, type -> value
+        """
+
+        # run the appropriate method for every dict item depending on value type
+        for (ch, p, typ), v in prop_params.items():
+            if typ == "I32":
+                self.SetProperty_i32(p, ch, v)
+            elif typ == "I64":
+                self.SetProperty_i64(p, ch, v)
+            elif typ == "Str":
+                self.SetProperty_s(p, ch, v)
+            else:
+                raise ValueError(f"Property type not handled: {typ}, valid types are I32/I64/Str")
+
+        logging.debug("parameter file device configuration property changes applied.")
+
+    @staticmethod
+    def parse_tsv_config(filestream):
+        """
+        Parse a tab-separated value (.mcs2.tsv) file in the following format:
+            (file header) channel_index     property    value   type   # description
+            channel_index can be a (Ch) number starting from 0 this also refers to a defined channel type
+            property is a hexadecimal code representing a specific property
+            value is a hexadecimal code representing independent property ﬂags
+            type is the type of the value used for this property (I32/I64/Str)
+        filestream (File): opened filestream
+        return:
+            prop_params_i32 (dict (int, int, str) -> int/str): channel index, property code, type -> value
+        """
+
+        # MCS2 property types dictionary:
+        prop_params = {}
+
+        # read the parameters "database" the file
+        for readline in filestream:
+            # comment or empty line?
+            mc = re.match(r"\s*(#|$)", readline)
+
+            if mc:
+                logging.debug("Comment line skipped: '%s'", readline.rstrip("\n\r"))
+                continue
+            # check if the read line matches the expected line format
+            m = re.match(
+                r"(?P<num>[0-9]+)\t(?P<code>0x[0-9A-F]+)\t(?P<value>(.*))\t(?P<type>(I32|I64|Str))\s*(#.*)?$",
+                readline,
+                re.IGNORECASE)
+
+            if not m:
+                # read line is not in the correct format
+                raise ValueError("Failed to parse line, unexpected line format '%s'" % readline.rstrip("\n\r"))
+
+            # assign the readline values to the appropriate variables
+            ch_num = int(m.group("num"))
+            prop_code = int(m.group("code"), 0)
+            typ = m.group("type")
+
```

Comment:
In this case when m is None the code will still try m.group which should not be possible since m is None.

Reply:
Good point, but these line are positioned above this block:

            if not m:
                # read line is not in the correct format
                raise ValueError("Failed to parse line, unexpected line format '%s'" % readline.rstrip("\n\r"))

Therefore, when m is None it is handled before assigning m.group to these variables.